major version release: 1.0.0

vahidm1993 opened the pull request:
Summary of pull request to merge in new features developed at Agile Robots AG.

* Not supporting Python 2 anymore
* Run this state
* Only run this state
* Add singleton pattern
* Add new hooks before and after running each state
* Add new memory profiling test to assert the memory leak during running sequential & concurrency state machines
* Update gaphas to 2.1.2
* Update libsass to the latest version of dart sass
* Replace gtkmvc3 with two separated mvc and observer patterns
* Run this state
* Removing the dead codes & comments

franzlst commented:
Sorry for not coming back to you so far.

I have already looked a quite some code, which looked good.

This morning, I wanted to install RAFCON, which turned out to be very hard now. `setup.py` now take sits requirements from [`requirements.txt`](https://github.com/AgileRobotsAG/RAFCON/blob/feat/final_major_one_version/requirements.txt). In there, there are three big issues:
* All version are fixed down to the patch version. Instead of allowing a certain version range (e.g. `pylint^=2.0.0`), you enforce a very specific version (e.g. `pylint==2.11.1`).
* You require very recent versions of packages that are hard to compile. For example, `PyGObject==3.40.1` or `pycairo==1.20.1`. Even OpenSUSE15.3 only comes with pycairo<1.20
* There are several packages listed, which I thing are not required:
  * matplotlib==2.1.1
  * monitoring==0.9.12 (?)
  * numpy==1.19.5 (?)
  * objgraph==3.5.0 (?)
  * pandas==1.1.5
  * PyOpenGL==3.1.5
  * PyYAML==6.0 (dependency of yaml_configuration)
  * rst2pdf==0.98
  * tf==1.0.0
  * Twisted==21.7.0

Other packages are probably only needed to run tests:
* profiling==0.1.3
* pykeyboard==0.1.2
* pymouse==1.0

Or create the documentation:
  * sphinx==4.3.0

If you want people to install RAFCON, all these points need to be fixed.

franzlst commented:
`platformdirs` should be restricted to <= 2.4.0, otherwise it's not compatible with Python 3.6. In theory, pip should know this and install v2.4.0, but in my case, it did not.

vahidm1993 commented:
> `platformdirs` should be restricted to <= 2.4.0, otherwise it's not compatible with Python 3.6. In theory, pip should know this and install v2.4.0, but in my case, it did not.

I do not see 'platformdirs' in 'requirements.txt' or any other places.

franzlst commented:
It's likely a dependeny of one the requirements. When you restrict the version the requirements.txt, it should have precedence over "sub-requirements"

franzlst commented:
It's [a dependency oy pylint](https://github.com/PyCQA/pylint/blob/main/setup.cfg#L47)

vahidm1993 commented:
matplotlib: execution_log.py -> log_to_ganttplot
monitoring: used in tests, move to the test requirements
numpy: vividict.py
objgraph: move to the test requirements
pandas: execution_log.py -> the function is going to be used in a test, move to the test requirements
rst2pdf: needed for generating documents, moved to the setup requirements
panda: execution_log.py -> log_to_DataFrame -> the function is going to be used in a test, move to the test requirements
tf: init_ros_node state machine
Twisted: core/start.py

franzlst commented:
> numpy: vividict.py

`Vividict` does not require numpy. It only supports numpy. there is a try/except block around `import numpy`. It should be up to the user to decide if he/she requires numpy. For running RAFCON, it's not required.

> Twisted: core/start.py

Same here. Twisted is not required, but supported. `def reactor_required` checks, if the user makes use of twisted and only then imports twisted. So twisted also here comes from additional _user_ requirements.

> tf: init_ros_node state machine

RAFCON does not need ROS. There might be an example state machine that uses ROS, but for this, no dependencies should be installed.

> matplotlib: execution_log.py -> log_to_ganttplot

I think log_to_ganttplot is never called directly. I would suggest to [create an extra in setup.py](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies) named "execution_log" or so, for which additional dependencies for the execution log are installed. Matplotlib is not a lightweight dependency.

vahidm1993 commented:
RIght, I wanted to remove 'log_to_ganttplot' & 'log_to_DataFrame' but @sebastian-brunner proposed keeping them & writing test cases for them, so they are called at least in one place. So, I move their dependencies to the test.

franzlst reviewed (commented):
I'm not yet fully through with the pull request, but have looked at 187 of 441 changed files. I left several comments, so that you do not run out of tasks ;-)

vahidm1993 commented on source/rafcon/gui/start.py:
```diff
@@ -330,6 +373,19 @@ def idle_handler(*args):
 
 
+def create_splash_screen():
+    if is_custom_design_enabled:
```
@sebastian-brunner You wrote the custom design. Is there any point for clarification or shall we refactor it?

sebastian-brunner replied:
Using the fallback value during config access is much nicer yes. @vahidm1993 : could you please update it accordingly?

vahidm1993 commented on source/rafcon/gui/design_config.py:
```diff
@@ -0,0 +1,69 @@
+global_design_config = DesignConfig(logger)
+
+
+def is_custom_design_enabled():
```
@sebastian-brunner Also here

sebastian-brunner replied:
Also this is a good idea by Franz. @vahidm1993 : could you change this accordingly?

vahidm1993 commented on source/rafcon/gui/mygaphas/guide.py:
```diff
@@ -13,68 +13,14 @@
 
 
 class GuidedStateMixin(GuideMixin):
-
```
Right, reverted.

vahidm1993 commented on source/rafcon/gui/mygaphas/tools.py:
```diff
@@ -121,22 +115,6 @@ def __init__(self, view=None, buttons=(1,)):
         self._move_name_v = False
         self._old_selection = None
 
-    def movable_items(self):
```
The function is being called during moving a state. I reverted the code.

vahidm1993 commented:
> Thank you adapting this!
> 
> * why is pytest and pytest-runner listed as setup_requirements?
> * I think you sometimes restrict now the _maximum_ version too much. E.g. `gaphas>=2.1.0,<=2.1.2`: if now gaphas is patched and release 2.1.3 is made, you do not allow it. Therefore, I would suggest that you always allow for patches and oftentimes even allow for minor releases (e.g. pylint). You often so so already, e.g. `astroid>=2.0.0,<3.0.0`

* I removed pytest from the setup_requirements. pytest-runner was already there and was not added in this MR, but by looking at the package definition, it adds the testing capability to setup.py.
* I fixed the gapha requirement.

franzlst commented on source/rafcon/gui/start.py:
```diff
@@ -199,14 +209,46 @@ def setup_argument_parser():
+        # the constant has to be overwritten here,
+        # as the singleton was already loaded that hus the wrong INTERFACE_FONT was chosen
+        from rafcon.gui.utils import constants
+        constants.INTERFACE_FONT = global_design_config.get_config_value("PRIMARY_FONT")
```
Why is this done only for design passed via the command line and not those set through the env var? I think th if and else branch can be merged.

franzlst commented on source/rafcon/gui/utils/constants.py:
```diff
@@ -28,6 +29,10 @@ def get_glade_path(glade_file):
 INTERFACE_FONT = "Source Sans Pro"
+if is_custom_design_enabled():
+    # this statement has no effect if the design_config is loaded after this module is imported the first time
+    INTERFACE_FONT = global_design_config.get_config_value("PRIMARY_FONT")
```
Shouldn't all this be possible in one line of code?

`INTERFACE_FONT = global_design_config.get_config_value("PRIMARY_FONT", "Source Sans Pro")`

franzlst commented on source/rafcon/gui/utils/splash_screen.py:
```diff
@@ -45,18 +46,25 @@ def __init__(self, width=530, height=350, contains_image=False):
-        self.label.set_yalign(0.5)
-        main_vbox.pack_start(self.label, False, True, 10)
-        main_vbox.set_spacing(0)
+        if is_custom_design_enabled() and not global_design_config.get_config_value("SPLASH_SCREEN_SHOW_TEXT", True):
```
Is this check for custom design really needed? And if so, what for?

vahidm1993 replied:
No, it wasn't needed.

franzlst reviewed (commented):
Some more minor comments. I'm now at >50% viewed files :-)

Your are fixing faster than I can spot issues ;-)

franzlst commented:
>     * I removed pytest from the setup_requirements. pytest-runner was already there and was not added in this MR, but by looking at the package definition, it adds the testing capability to setup.py.

Perfect!

>     * I fixed the gapha requirement.

Where? Did you forget to commit this?

vahidm1993 commented:
> > ```
> > * I removed pytest from the setup_requirements. pytest-runner was already there and was not added in this MR, but by looking at the package definition, it adds the testing capability to setup.py.
> > ```
> 
> Perfect!
> 
> > ```
> > * I fixed the gapha requirement.
> > ```
> 
> Where? Did you forget to commit this?

I currently set a range as follows to install 2.1.x:

gaphas>=2.1.0,<2.2.0

I had to fix the minor package to '1' since Rafcon only works with this specific minor version. Downgrading/upgrading this minor version breaks the code as some classes (e.g., projection) are moved/changed in those minor releases.

franzlst commented:
Ah I see!

franzlst commented on source/rafcon/gui/design_config.yaml:
```diff
@@ -0,0 +1,16 @@
+TYPE: DESIGN_CONFIG
+USE_CUSTOM_DESIGN: False
+FONTS_FOLDER: $RELATIVE_PATH{../../../share/fonts}
+PRIMARY_FONT: "Quaerite Regnum Dei"
```
I think this should you Source Sans Pro, as this is shipped with RAFCON

franzlst commented on source/rafcon/gui/design_config.yaml:
```diff
@@ -0,0 +1,16 @@
+TYPE: DESIGN_CONFIG
+USE_CUSTOM_DESIGN: False
```
Is this option still required?

vahidm1993 replied:
yes, I see it is still used in 3 other places in gui/config.py

franzlst commented on source/rafcon/gui/design_config.yaml:
```diff
@@ -0,0 +1,16 @@
+THEME_NAME: RAFCON
+SOURCE_VIEW_FOLDER: $RELATIVE_PATH{../../../share/gtksourceview-3.0}
+# overwrites gui_config.SOURCE_EDITOR_STYLE
+SOURCE_VIEW_THEME: rafcon-agile
```
Just `rafcon`?

vahidm1993 replied:
any suggestion?

sebastian-brunner replied:
yes just rafcon

franzlst commented on source/rafcon/gui/start.py:
```diff
@@ -199,14 +209,46 @@ def setup_argument_parser():
+        from rafcon.gui.utils import constants
+        constants.INTERFACE_FONT = global_design_config.get_config_value("PRIMARY_FONT")
+    elif os.environ.get('RAFCON_CUSTOM_DESIGN'):
+        design_config = os.environ.get('RAFCON_CUSTOM_DESIGN')
```
I think this is the only line that differs between `if` and `else` branch. The rest can be moved out of the branch and is currently duplicate code

franzlst commented on source/rafcon/gui/utils/constants.py:
```diff
@@ -19,6 +19,7 @@
 from gi.repository import Gdk
 
 from rafcon.utils.constants import RAFCON_TEMP_PATH_BASE
+from rafcon.gui.design_config import global_design_config, is_custom_design_enabled
```
`is_custom_design_enabled` is nor more needed, right?

franzlst commented on source/rafcon/gui/utils/splash_screen.py:
```diff
@@ -45,41 +46,55 @@ def __init__(self, width=530, height=350, contains_image=False):
-            images.append(resource_filename(rafcon.gui.__name__, os.path.join("assets", "splashscreens", image_filename)))
+    def get_images(self):
+        images = list()
+        if is_custom_design_enabled():
```
I think this differentiation is no longer needed, as now also the default design config defines `SPLASH_SCREEN_FOLDER`

franzlst commented on source/rafcon/utils/installation.py:
```diff
@@ -53,6 +54,11 @@ def install_fonts(restart=False):
     font_names_to_be_installed = ["SourceSansPro", "FontAwesome5Free", "RAFCON"]
 
+    custom_fonts = list()  # support a list in case we also support secondary or tertiary fonts
+    if is_custom_design_enabled():
```
should also be obsolete, as the default config defines this option

franzlst commented on source/rafcon/utils/installation.py:
```diff
@@ -102,3 +109,13 @@ def install_locally_required_files():
         except IOError as e:
             logger.error("Could not copy '{}' files: {}".format(folder, str(e)))
+
+    if is_custom_design_enabled():
```
might also be required in any case

franzlst commented on tests/core/test_gvm.py:
```diff
@@ -140,15 +140,15 @@ def test_type_check(caplog):
     key_start_a = gvm.get_all_keys_starting_with('a')
-    assert (len(gvm.get_all_keys()) == 4)
-    assert (len(key_start_a) == 3)
+    assert (len(gvm.get_all_keys()) == 33)
```
I guess the change in number arises from the new singleton concept. In my opinion, there needs to be a way to reset a singleton, and this should be done before every test. Otherwise, this test is obsolete and cannot be run independent from other tests.

franzlst reviewed (changes requested):
Yay, I'm done!

It's good to say that most of the new features have unit tests!

I left some comment. Several of them refer to the design config, I think the code can further be simplified here.

I also have one comment regrading the new singleton concept. Please think about this!

vahidm1993 commented on source/rafcon/design_patterns/singleton.py:
```diff
@@ -16,6 +16,14 @@ def instance(self, *args, **kwargs):
             self._instance = self._cls(*args, **kwargs)
         return self._instance
 
+    def reset_and_instance(self, *args, **kwargs):
```
Do you have another solution?

franzlst replied:
I think that singletons are not a very good pattern, as you can see in this case.

But the singletons are new there, so I propose to add a `reset()` method to the GVM that removes all variables. Instead of calling then `reset_and_instance()`, you first call `instance()` and then `reset()` on the instance.

The question is if not other singletons also hold some state which should be reset. The `ExecutionEngine` probably holds a state machine, etc.

sebastian-brunner reviewed (approved):
All changes were adressed by @vahidm1993

sebastian-brunner commented:
Thx for the review and the fixes.

sebastian-brunner commented:
This one was merged already.
